### PR TITLE
introduce tags support to VMs

### PR DESCRIPTION
**Pull request by eslutsky**
## Please describe the change you are making
introduce tags support for vms

### CODE REVIEW

File: client_tag_create.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
+// Code generated automatically using go:generate. DO NOT EDIT.
```
Comment (janosdebugs) [inline client_tag_create.go:1]: This code is not auto-generated, please remove this line to avoid confusion.

File: client_tag.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+import (
+	ovirtsdk4 "github.com/ovirt/go-ovirt"
+)
+
+//go:generate go run scripts/rest.go -i "Tag" -n "tag"
+
+// TagClient describes the functions related to oVirt tags.
+type TagClient interface {
+	// GetTag returns a single network based on its ID.
```
Comment (janosdebugs) [inline client_tag.go:11]: Description is incorrect.

File: mock_tag_create.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+	defer m.lock.Unlock()
+
+	tag := m.createTag(name, description)
+
+	result = tag
+
+	return
+}
+
+func (m *mockClient) createTag(
```
Comment (janosdebugs) [inline mock_tag_create.go:16]: What is the purpose of this separate function?

File: mock_tag_remove.go [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+func (m *mockClient) RemoveTag(id string, _ ...RetryStrategy) (err error) {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	if _, ok := m.tags[id]; !ok {
+		return newError(ENotFound, "Tag with ID %s not found", id)
+	}
+
+	delete(m.tags, id)
```
Comment (janosdebugs) [inline mock_tag_remove.go:11]: The tags need to be removed from the VMs too in this case.

File: client_vm_add_tag.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package ovirtclient
+
+import (
+	"fmt"
+
+	ovirtsdk "github.com/ovirt/go-ovirt"
+)
+
+func (o *oVirtClient) AddTagToVM(id string, tagID string, retries ...RetryStrategy) (err error) {
```
Comment (janosdebugs) [inline client_vm_add_tag.go:9]: Please add a convenience function to the VM object to call this function.

**Review by janosdebugs — CHANGES_REQUESTED** [review]
A few small changes please.

File: client_vm.go [hunk 6]
```diff
@@ -612,6 +635,22 @@ func (v *vm) Name() string {
 	return v.name
 }
 
+func (v *vm) TagIDs() []string {
+	return v.tagIDs
+}
+
+func (v *vm) Tags(retries ...RetryStrategy) ([]Tag, error) {
```
[the remark this answers is not in the document]
Comment (eslutsky) [reply client_vm.go:644]: done

File: client_vm_add_tag.go (hunk 5 above)
Comment (eslutsky) [reply client_vm_add_tag.go:9]: done

File: mock_tag_remove.go (hunk 4 above)
Comment (eslutsky) [reply mock_tag_remove.go:22]: done

File: client_tag_create.go (hunk 1 above)
Comment (eslutsky) [reply client_tag_create.go]: done

File: mock_tag_create.go (hunk 3 above)
Comment (eslutsky) [reply mock_tag_create.go]: removed

**Review by janosdebugs — APPROVED** [review]
LGTM

**Comment by janosdebugs** [comment]
@eslutsky you are good to go, just rebase and merge.

**Review by janosdebugs — APPROVED** [review]

**Review by janosdebugs — APPROVED** [review]

**Pull request merged**